lc0/src/neural/network_blas_cl.h

```diff
@@ -0,0 +1,176 @@
+Network::initialize::channels -> ConvBlock.biases.size()
+conv_weights             -> ConvBlock.weights
+conv_biases              -> ConvBlock.biases
+batchnorm_means          -> ConvBlock.bn_means
+batchnorm_stddivs        -> ConvBlock.bn_stddivs
+conv_pol_w               -> Weights.policy.weights
+conv_pol_b               -> Weights.policy.biases
+bn_pol_w1                -> Weights.policy.bn_means,
+bn_pol_w2                -> Weights.policy.bn_stddivs // ??? not very useful oldnames
+NUM_POLICY_INPUT_PLANES  -> Weights.policy.bn_means.size()
+ip_pol_w                 -> Weights.ip_pol_w
+ip_pol_b                 -> Weights.ip_pol_b
+get_num_output_policy()  -> Weights.ip_pol_b.size()
+conv_val_w               -> Weights.value.weights
+conv_val_b               -> Weights.value.biases
+bn_val_w1                -> Weights.value.bn_means
+bn_val_w2                -> Weights.value.bn_stddivs
+NUM_VALUE_INPUT_PLANES   -> Weights.value.bn_means.size()
+{ip1_val_w,ip1_val_b,ip2_val_w,ip2_val_b} -> Weights.{}
+NUM_VALUE_CHANNELS       -> Weights.ip1_val_b.size()
+
+*/
+
+/* This file defines the BlasNetwork and OpenCLNetwork classes.
+
+LeelaZero's original OpenCL backend was written using a few pieces of the blas
+backend (including initialization) as helper code; meanwhile, the OpenCL
+selfcheck also *calls* the CPU implementation to verify the GPU computations,
+which means that it's probably more efficient if the CPU class and GPU class
+share network weights, so the OpenCLNetwork just directly inherits from BlasNetwork.
+Protected members are those which the OpenCL implementation uses, while private
+members are those used exclusively by the CPU implementation.
+*/
+
+#pragma once
+
+#include "factory.h" // network.h, optionsdict.h
+#include "blas_config.h"
+#include "opencl/OpenCLScheduler.h"
+
+namespace lczero {
+
+class BlasNetwork;
+
+class BlasNetworkComputation : public NetworkComputation {
+
+ public:
+  BlasNetworkComputation(BlasNetwork* network) : network(network) {}
+
+  void AddInput(InputPlanes&& input) override {
+    inputs.emplace_back(input);
+  }
+
+  int GetBatchSize() const override {
+    return inputs.size();
+  }
+
+  float GetQVal(int sample) const override {
+    return output_values[sample];
+  }
+
+  float GetPVal(int sample, int move_id) const override {
+    return output_policies[sample][move_id];
+  }
+
+  void ComputeBlocking() override;
+
+ private:
+  std::vector<InputPlanes> inputs;
+  std::vector<float>   output_values;
+  std::vector<std::vector<float>> output_policies;
+  BlasNetwork* network;
+};
+
+class BlasNetwork : public Network {
+
+ public:
+  BlasNetwork(const Weights& weights, const OptionsDict& options);
+
+  std::unique_ptr<NetworkComputation> NewComputation() override {
+    return std::make_unique<BlasNetworkComputation>(this);
+  }
+
+  std::pair<float, std::vector<float>> evaluate(InputPlanes& input);
+  // this is the main function used by outside code
```

Comment:
What does this comment relate to?

Reply:
evaluate(), I guess it should go above the function not below